pink/barriers/barrier.py

```diff
@@ -0,0 +1,221 @@
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Value of the barrier function :math:`\boldsymbol{h}(\boldsymbol{q})`.
+        """  # noqa: E501
+
+    @abc.abstractmethod
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the Jacobian matrix of the barrier function.
+
+        The Jacobian matrix
+        :math:`\frac{\partial \boldsymbol{h}}{\partial \boldsymbol{q}}(\boldsymbol{q})`
+        of the barrier function with respect to the configuration variables is
+        required for the computation of the barrier condition.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Jacobian matrix
+            :math:`\frac{\partial \boldsymbol{h}}{\partial \boldsymbol{q}}(\boldsymbol{q})`.
+        """  # noqa: E501
+
+    def compute_safe_policy(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the safe backup control policy.
+
+        The safe backup control policy
+        :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})` is a joint
+        velocity vector that can be used as a regularization term in the
+        optimization problem to ensure safety.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+
+        Returns:
+            Safe backup joint velocities
+                :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})`.
+        """
+        return np.zeros(configuration.model.nv)
+
+    def compute_qp_objective(
+        self,
+        configuration: Configuration,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        r"""Compute the quadratic objective function for the barrier-based QP.
+
+        The quadratic objective function includes a regularization term
+        based on the safe backup policy:
+
+        .. math::
+
+            \gamma(\boldsymbol{q})\left\| \dot{\boldsymbol{q}}-
+            \dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})\right\|^{2}
+
+        where :math:`\gamma(\boldsymbol{q})` is a configuration-dependent
+        weight and :math:`\dot{\boldsymbol{q}}_{safe}(\boldsymbol{q})`
+        is the safe backup policy.
+
+        Args:
+            configuration: Robot configuration :math:`\boldsymbol{q}`.
+            dt: Time step for discrete-time implementation. Defaults to 1e-3.
+
+        Returns:
+            Tuple containing the quadratic objective matrix (H) and linear
+                objective vector (c).
+        """
+        jac = self.compute_jacobian(configuration)
+        H = np.zeros((configuration.model.nv, configuration.model.nv))
+        # c = 1e-3 * np.linalg.norm(jac, axis=0) / dt
+        c = np.zeros(configuration.model.nv)
+
+        if self.r > 1e-6:
+            self.safe_policy = self.compute_safe_policy(configuration)
+
+            H += (
+                self.r
+                / (np.linalg.norm(jac) ** 2)
```

Comment:
What if the norm of the (user-supplied) Jacobian is close to zero?

According to NumPy's doc, for matrices this expression will compute the Frobenius norm. Is it fine performance-wise to compute a full norm here? (For instance, how long does it take on your machine to compute that part, relative to the time it takes to compute the barrier and barrier Jacobian?)

Reply:
We would investigate the performance issues and provide the results here.

As for norm close to zero, this is exactly why this term is introduced at the first place. According to the [original paper](https://arxiv.org/pdf/2404.12329), when the jacobian is close to 0, this term enforces system to choose safe displacement, which does not violate a barrier constraint.